ICP-8617, ICP-8572 Update supported modes for LUX KONOz thermostat

dkirker opened the pull request.

dkirker committed:
ICP-8617, ICP-8572 Update supported modes for LUX KONOz thermostat

dkirker committed:
Restrict behavior for setThermostatMode and setThermostatFanMode

dkirker committed:
Remove duplicate "nextState"

dkirker committed:
Add major updates for dynamic thermostat modes (not complete yet)

tpmanley commented on devicetypes/smartthings/zigbee-thermostat.src/zigbee-thermostat.groovy:
```diff
@@ -264,26 +304,53 @@ def getPowerSource(value) {
+def switchMode() {
+	def currentMode = device.currentValue("thermostatMode")
+	def modeOrder = state.supportedThermostatModes
+log.debug "switchMode() $currentMode $modeOrder"
```
nit: indentation on the log statements

dkirker replied:
Yeah, these will go away shortly. The PR isn't 100% done yet.

Actually, Bob shared some code for changes made to multi-attr tiles some time ago that should make this function irrelevant. :smile:

dkirker replied:
With some updated tile stuff from @bflorian this function is no longer needed, so it is gone!

dkirker committed:
Update ST Classic support
Better handle supported modes (state.supportedModes seemed to not work), get ready for dynamic setpoint ranges

dkirker commented:
@tpmanley  @greens  The PR has been updated with better Classic support. I did start adding some code that will be used for setpoint range support. The thermostat allows you to configure these values. Right now I have defined the values statically like in the metadata.

tpmanley commented on devicetypes/smartthings/zigbee-thermostat.src/zigbee-thermostat.groovy:
```diff
@@ -35,72 +35,87 @@ metadata {
 		capability "Sensor"
 
-		command "lowerHeatingSetpoint"
+		/*command "lowerHeatingSetpoint"
```
Any reason not to remove these altogether?

tpmanley commented on devicetypes/smartthings/zigbee-thermostat.src/zigbee-thermostat.groovy:
```diff
@@ -35,72 +35,87 @@ metadata {
+				debouncePeriod: 1500,
+				range: "device.heatingSetpointRange",
+				width: 2, height: 2)
+				{
```
`{` should go on previous line

tpmanley commented on devicetypes/smartthings/zigbee-thermostat.src/zigbee-thermostat.groovy:
```diff
@@ -35,72 +35,87 @@ metadata {
+				debouncePeriod: 1500,
+				range: "device.coolingSetpointRange",
+				width: 2, height: 2)
+				{
```
`{` should go on previous line

tpmanley commented on devicetypes/smartthings/zigbee-thermostat.src/zigbee-thermostat.groovy:
```diff
@@ -35,72 +35,87 @@ metadata {
+		controlTile("thermostatFanMode", "device.thermostatFanMode", "enum", width: 2 , height: 2, supportedStates: "device.supportedThermostatFanModes") {
+			state "auto", action: "setThermostatFanMode", label: 'Auto', icon: "st.thermostat.fan-auto"
+			state "on",	action: "setThermostatFanMode", label: 'On', icon: "st.thermostat.fan-on"
+			//state "circulate", action: "setThermostatFanMode", label: 'Circulate', icon: "st.thermostat.fan-circulate"
```
Remove?

tpmanley commented on devicetypes/smartthings/zigbee-thermostat.src/zigbee-thermostat.groovy:
```diff
@@ -124,7 +137,7 @@ def parse(String description) {
 		log.warn "Unexpected event: ${map}"
 	}
-	log.debug "Description ${description} parsed to ${result}"
+	//log.debug "Description ${description} parsed to ${result}"
```
Remove altogether? Or leave it uncommented - I don't think it hurts anything and can be useful

tpmanley commented on devicetypes/smartthings/zigbee-thermostat.src/zigbee-thermostat.groovy:
```diff
@@ -138,47 +151,80 @@ private parseAttrMessage(description) {
+					map.value = binValue[2] == "1" ? "fan only" : "idle"
+				}
+			} else if (it.attribute == THERMOSTAT_SYSTEM_CONFIG) {
+				def binValue = extendString(bin(hexToInt(it.value)), 16, '0').reverse()
```
I think it would be more clear to convert to a number instead of leaving it as a string

greens replied:
yeah i think bitmasking is the better way to go here too

dkirker replied:
Finishing that up; same for THERMOSTAT_RUNNING_STATE

dkirker replied:
Done!

greens commented on devicetypes/smartthings/zigbee-thermostat.src/zigbee-thermostat.groovy:
```diff
@@ -138,47 +151,80 @@ private parseAttrMessage(description) {
+				def cooling = binValue[0..1].reverse()
+				def heating = binValue[2..3].reverse()
+				def heatingType = binValue[4]
+				def heatingFuel = binValue[5]
```
this is unused, right?

dkirker replied:
Yeah. I had included it since it was a representation of the data, however, best to not waste the extra memory and instructions.

greens commented on devicetypes/smartthings/zigbee-thermostat.src/zigbee-thermostat.groovy:
```diff
@@ -264,26 +317,41 @@ def getPowerSource(value) {
+			case "auto":
+				fanAuto()
+				break
+		}
```
you could add a debug message here too

greens commented on devicetypes/smartthings/zigbee-thermostat.src/zigbee-thermostat.groovy:
```diff
@@ -377,19 +455,38 @@ private extendString(str, size, character) {
+}
+def getMaxSetpointIndex() {
+	1
+}
```
both of these seem to be unused

dkirker committed:
Code formatting cleanup

dkirker committed:
Refactor supportedThermostatModes and supportedThermostatFanModes

dkirker committed:
A little more cleanup

dkirker committed:
Remove unused functions

dkirker commented:
Alright @tpmanley @greens I did a bit more refactoring. I had been trying out code for CONTROL_SEQUENCE_OF_OPERATION (from the Zen Thermostat DTH) which is a mandatory attribute, however I opted to stick with THERMOSTAT_SYSTEM_CONFIG for determining supportedThermostatModes since the KONOZ does support THERMOSTAT_SYSTEM_CONFIG. I decided to leave the CONTROL_SEQUENCE_OF_OPERATION code in there so if another thermostat is added to this DTH that doesn't support THERMOSTAT_SYSTEM_CONFIG that CONTROL_SEQUENCE_OF_OPERATION code can be used. That, and if someone copies parts from this DTH there's a bit more code to prevent influencing bad behavior.

bretschuring commented on devicetypes/smartthings/zigbee-thermostat.src/zigbee-thermostat.groovy:
```diff
@@ -109,92 +116,181 @@ metadata {
 	return result
 }
 
 private parseAttrMessage(description) {
```
This function is a little long. It might make sense to break out the individual attribute handling into separate functions.

bretschuring commented on devicetypes/smartthings/zigbee-thermostat.src/zigbee-thermostat.groovy:
```diff
@@ -109,92 +116,181 @@ metadata {
+				def heatingType = (intValue & 0b00010000) >>> 4
+				def supportedModes = ["off"]
+
+				if (cooling != 0x03) {
```
It would be nice to make some of these numbers constants with a descriptive name. Then it will be more clear what is being checked.

dkirker replied:
I'll look into doing another PR that does some cleanup. Thanks!
